File: leetcode/stacks/20-valid-parentheses/c/parentheses_validator.c
#include <stdbool.h>
#include <stdlib.h>

#include "stack.h"

bool isValid(char *s)
{
  // Variables
  char *currentChar = s;
  bool isClosingBracketMatching = true;
  uint32_t count = 0;

  // Create the stack
  void *parenthesesStack = stackCreate();

  // Loop through individual characters
  while (*currentChar != '\0')
  {
    if (*currentChar == '(' || *currentChar == '{' || *currentChar == '[')
    {
      char *openingBracket = calloc(2, sizeof(char));
      openingBracket[0] = *currentChar;
      stackPush(parenthesesStack, openingBracket);
      ++count;
    }
    else if (*currentChar == ')')
    {
      char *lastOpeningBracket = stackPop(parenthesesStack);
      if (lastOpeningBracket[0] != '(')
      {
        isClosingBracketMatching = false;
        break;
      }
      ++count;
    }
    else if (*currentChar == '}')
    {
      char *lastOpeningBracket = stackPop(parenthesesStack);
      if (lastOpeningBracket[0] != '{')
      {
        isClosingBracketMatching = false;
        break;
      }
      ++count;
    }
    else if (*currentChar == ']')
    {
      char *lastOpeningBracket = stackPop(parenthesesStack);
      if (lastOpeningBracket[0] != '[')
      {
        isClosingBracketMatching = false;
        break;
      }
      ++count;
    }
    currentChar++;
  }

  // Return the result
  bool result = IS_STACK_EMPTY(parenthesesStack) && count > 0 ? isClosingBracketMatching : false;

  // Memory cleanup
  while (!IS_STACK_EMPTY(parenthesesStack))
    free(stackPop(parenthesesStack));

  stackDelete(parenthesesStack);

  return result;
}

File: leetcode/stacks/20-valid-parentheses/js/parentheses-validator.js
/**
 * @param {string} s
 * @return {boolean}
 */
function isValid(s) {
  const length = s.length;

  /** @type {string[]} */
  const stack = [];

  /** @type {string[][]} */
  const [openingBrackets, closingBrackets] = [
    ["(", "{", "["],
    [")", "}", "]"],
  ];

  /** @type {boolean[]} */
  let [doParensMatch, containsInvalidChars] = [true, false];

  for (let i = 0; i < length; i++) {
    const currentChar = s[i];

    if (openingBrackets.includes(currentChar)) {
      stack.push(currentChar);
    } else if (closingBrackets.includes(currentChar)) {
      if (stack.length === 0) {
        doParensMatch = false;
        break;
      } else {
        const popped = stack.pop();

        if (
          (currentChar === ")" && popped !== "(") ||
          (currentChar === "}" && popped !== "{") ||
          (currentChar === "]" && popped !== "[")
        ) {
          doParensMatch = false;
          break;
        }
      }
    } else {
      containsInvalidChars = true;
    }
  }

  const result = !containsInvalidChars && doParensMatch && stack.length === 0;
  return result;
}
